idaes/core/util/model_diagnostics.py

```diff
@@ -17,29 +17,987 @@
 
 __author__ = "Alexander Dowling, Douglas Allan, Andrew Lee"
 
-
 from operator import itemgetter
+from sys import stdout
 
-import pyomo.environ as pyo
-from pyomo.core.expr.visitor import identify_variables
-from pyomo.contrib.pynumero.interfaces.pyomo_nlp import PyomoNLP
-from pyomo.contrib.pynumero.asl import AmplInterface
-from pyomo.core.base.block import _BlockData
 import numpy as np
 from scipy.linalg import svd
 from scipy.sparse.linalg import svds, norm
 from scipy.sparse import issparse, find
 
+from pyomo.environ import (
+    Binary,
+    Block,
+    check_optimal_termination,
+    ConcreteModel,
+    Constraint,
+    Objective,
+    Param,
+    Set,
+    SolverFactory,
+    value,
+    Var,
+)
+from pyomo.core.base.block import _BlockData
+from pyomo.common.collections import ComponentSet
+from pyomo.common.config import ConfigDict, ConfigValue, document_kwargs_from_configdict
+from pyomo.util.check_units import identify_inconsistent_units
+from pyomo.contrib.incidence_analysis import IncidenceGraphInterface
+from pyomo.core.expr.visitor import identify_variables
+from pyomo.contrib.pynumero.interfaces.pyomo_nlp import PyomoNLP
+from pyomo.contrib.pynumero.asl import AmplInterface
+from pyomo.common.deprecation import deprecation_warning
+
 from idaes.core.util.model_statistics import (
+    activated_blocks_set,
+    deactivated_blocks_set,
+    activated_equalities_set,
+    deactivated_equalities_set,
+    activated_inequalities_set,
+    deactivated_inequalities_set,
+    activated_objectives_set,
+    deactivated_objectives_set,
+    variables_in_activated_constraints_set,
+    variables_not_in_activated_constraints_set,
+    degrees_of_freedom,
     large_residuals_set,
     variables_near_bounds_set,
 )
-import idaes.core.util.scaling as iscale
+from idaes.core.util.scaling import (
+    get_jacobian,
+    extreme_jacobian_columns,
+    extreme_jacobian_rows,
+    extreme_jacobian_entries,
+    jacobian_cond,
+)
 import idaes.logger as idaeslog
 
 _log = idaeslog.getLogger(__name__)
 
 
+MAX_STR_LENGTH = 84
+TAB = " " * 4
+
+# TODO: Add suggested steps to cautions - how?
+
+CONFIG = ConfigDict()
+CONFIG.declare("model", ConfigValue(description="Pyomo model object to be diagnosed."))
```

Comment:
I've changed `model` to be a positional argument which is used to set a private `self._model` attribute (and updated the docs accordingly). I've also added `self.model` as a `@property` so that users can access (but not change) it easily as well as an explicit `self.set_model(model)` method to allow users to change the model they are looking at.

Reply:
Do we expect users to need to (a) retrieve and (b) change the model that their `DiagnosticsToolbox` instance is currently debugging?

Reply:
Maybe - being able to retrieve it lets you confirm what model (or Block) you are looking at.

As for changing the model, I could see it being useful for a few reasons, such as switching between scaled and unscaled version of the same model, or when trying to debug more complex hierarchical models whilst keeping your current config options (e.g. moving from `m.b` to `m`). That said, creating a new instance of the toolbox is also cheap so you could argue it is better to just make users create a new instance for clarity (although then you have multiple instances of the toolbox around which could be just as confusing).

Reply:
Changing to a "sub-model" with the same config settings is the most compelling reason to me. Although you could just create a new instance and use `**old_dtbox.config` as keyword args.

I like the `model` property, but at the moment I'm against letting users change the model. Letting the model change adds more mutable state that the user has to keep track of, and there is a pretty easy workaround (create a new instance). It will also make `DiagnosticsToolbox` more difficult to extend, because if we ever do any caching, we will have to invalidate all the caches.

Reply:
I've removed the `set_model` method.